source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+
+A value of type ``pulse`` is retrieved by calling one of the built-in pulse
+template functions. Note that each of these functions takes a type ``length``
+as a first argument, since pulses need to have a definite length. Using the
+hardware dependent ``dt`` unit is recommended, since the compiler may need to
+down-sample a higher precision pulse to physically realize it.
+
+.. code-block:: none
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   gaussian(length:l, complex[float[32]]:amp, length:center, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   sech(length:l, complex[float[32]]:amp, length:center, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // square_width is the width of the square pulse component
+   // sigma is the standard deviation of pulse
+   gaussian_square(length:l, complex[float[32]]:amp, length:center, length:square_width, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   // beta is the Y correction amplitude, see the DRAG paper
+   drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
+
+   // Define a boxcar (aka. constant) pulse of length l
+   boxcar(l:length)
+
+Play instruction
+~~~~~~~~~~~~~~~~
+
+Pulses are scheduled using the ``play`` instruction. Play instructions have two
+required parameters:
+
+- the channel on which to play the pulse
+- a value of type ``pulse`` representing the pulse envelope
+
+Optionally the ``play`` instruction can also take a third parameter: the frame
+to use for the pulse. If the frame is not specified then the corresponding frame
+for the channel will be used instead, as if ``frameof`` was called for that
+channel.
+
+.. code-block:: none
+
+   // Play a 3 sample pulse on qubit 0
+   play transmit(%0), [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // Play a gaussian on qubit 0 using qubit 1's frame
+   frame f1 = frameof(transmit(%1));
+   play transmit(%0), gaussian(...), f1;
+
+Capture Instruction
+-------------------
+
+Acquisition is scheduled by a ``capture`` instruction. This is a special
+``kernel`` function which is specified by a hardware vendor. The measurement
+process is difficult to describe generically due to the wide variety of
+hardware and measurement methods.
+
+The only required parameter is an ``channel``, ie. a channel returned from
+calling the ``receive`` function.
+
+The following are possible parameters that might be included:
+
+- A "duration" of type ``length``, if it cannot be inferred from other parameters
+- A "filter", which is dot product-ed with the measured IQ the distill the
+  result into a single IQ value
+- A "tag", which could be used to identify which branch of an if statement was
+  traversed
```

Comment:
I'm struggling to imagine a scenario where the measurement operation would need this info. Did you have something particular in mind?

Reply:
I believe the rationale was if there is data we could return in real-time to the circuit level that the experimentalist wishes to later extract and debug offline, like the full sampled signal. The measurement in question could be captured.

Reply:
Ya basically. I'm predicting that in the near term the accumulation of results will be done mostly as a side effect (ie. you call capture without assigning it to a variable) and the results just buffer somewhere outside the language. Then it would be useful to tag multiple measurements.

Or in the case of active reset: the `reset` statement doesn't return anything, but maybe you're interested in how many rounds of active reset you do on average.

In any case with the generic capture instruction if a hardware vendor doesn't want to implement this they don't need to.